docs/develop/core.rst

```diff
@@ -0,0 +1,393 @@
+
+Defining the same rule without a compile function is achieved by using the ``DEF_RULE_NC`` function
+and omitting the compile function argument:
+
+.. code-block:: c
+
+   DEF_RULE_NC(add1_stmt, and(2), tok(KW_ADD))
+
+The remaining arguments take on the same meaning.
+
+.. note::
+   The function ``DEF_RULE`` or ``DEF_RULE_NC`` takes other arguments. For an indepth understanding
+   on what parameters are supported, see ``py/grammar.h``.
+
+Changing the lexer
+~~~~~~~~~~~~~~~~~~
+
+Every rule defined in the grammar should have a token associated with it and defined in ``py/lexer.h``. 
+Add this token by editing the ``_mp_token_kind_t`` enum:
+
+.. code-block:: c
+   :emphasize-lines: 12
+
+   typedef enum _mp_token_kind_t {
+    ...
+    ...
+    P_TOKEN_KW_OR,
+    MP_TOKEN_KW_PASS,
+    MP_TOKEN_KW_RAISE,
+    MP_TOKEN_KW_RETURN,
+    MP_TOKEN_KW_TRY,
+    MP_TOKEN_KW_WHILE,
+    MP_TOKEN_KW_WITH,
+    MP_TOKEN_KW_YIELD,
+    MP_TOKEN_KW_ADD1,
+   } mp_token_kind_t;
+
+Since, we are adding a keyword, edit ``py/lexer.c`` to add the new keyword:
+
+.. code-block:: c
+   :emphasize-lines: 12
+
+   STATIC const char *const tok_kw[] = {
+    "False",
+    "None",
+    "True",
+    "__debug__",
+    "and",
+    "as",
+    "assert",
+    #if MICROPY_PY_ASYNC_AWAIT
+    "async",
+    "await",
+    #endif
+    "break",
+    "class",
+    "continue",
+    "def",
+    "del",
+    "elif",
+    "else",
+    "except",
+    "finally",
+    "for",
+    "from",
+    "global",
+    "if",
+    "import",
+    "in",
+    "is",
+    "lambda",
+    "nonlocal",
+    "not",
+    "or",
+    "pass",
+    "raise",
+    "return",
+    "try",
+    "while",
+    "with",
+    "yield",
+    "add1",
+   };
+
+Notice the keyword is named depending on what you want it to be. For consistency, maintain the
+naming standard accordingly.
+
+.. note::
+   The order of these keywords in ``py/lexer.h`` should match the order of tokens in the enum
+   defined in ``py/lexer.h``.
+
+Parsing
+~~~~~~~~
+
+The parser takes the tokens produced by the lexer converting them to an abstract syntax tree(AST) or
+parse tree. The implementation for the parser is defined in ``py/parse.c``. 
+
+The parser also maintains a table of constants for use in different aspects parsing, similar to what a `symbol 
+table <https://steemit.com/programming/@drifter1/writing-a-simple-compiler-on-my-own-symbol-table-basic-structure>`_ 
+does.
+
+Several optimizations like `constant folding <http://compileroptimizations.com/category/constant_folding.htm>`_ 
+on integers for all operations i.e logical, binary, unary, etc, optimizing enhancements on parenthesis
+around expressions are performed during this phase and optimizations on strings.
+
+Notable to the parser is that at this stage, some lonely statements like *docstrings* are discarded and not 
+accessible to the compiler. Even optimizations like `string interning <https://en.wikipedia.org/wiki/String_interning>`_ are 
+not applied to *docstrings*.
```

Comment:
"Notable" and "lonely" don't quite work in this paragraph, perhaps it might help to simplify to something like:

```suggestion
It's worth noting that *docstrings* are discarded and not 
accessible to the compiler. Even optimizations like `string interning <https://en.wikipedia.org/wiki/String_interning>`_ are 
not applied to *docstrings*.
```